lib/plugins/threaded-signing/plugin.c

```diff
@@ -188,59 +245,67 @@ threaded_openssl_sign_hash(sv_threaded_plugin_t *self, const signature_info_t *s
   SignedVideoReturnCode status = SV_UNKNOWN_FAILURE;
   g_mutex_lock(&self->mutex);
 
-  // If the signature has not yet been pulled, or even generated, a new signature cannot be
-  // generated without replacing it. Log in |nbr_of_unsigned_hashes| and move to done.
-  if (self->plugin_state != THREADED_SIGNING_WAITS_FOR_HASH_TO_SIGN) {
-    self->nbr_of_unsigned_hashes++;
-    status = SV_OK;
-    goto done;
+  if (!self->is_running) {
+    // Thread is not running. Go to catch_error and return status.
+    status = SV_EXTERNAL_ERROR;
+    goto catch_error;
   }
 
   // If no |self->signature_info| exists. Allocate necessary memory for it and copy the
   // |private_key| and |algo|.
   if (!self->signature_info) {
     self->signature_info = local_signature_info_create(signature_info);
-    if (!self->signature_info) goto catch_error;
+    if (!self->signature_info) {
+      // Failed in memory allocation.
+      status = SV_MEMORY;
+      goto catch_error;
+    }
   }
 
-  // Allocate memory for the |hash_to_sign| if necessary.
-  if (!self->hash_to_sign) {
-    self->hash_to_sign = calloc(1, signature_info->hash_size);
-    if (!self->hash_to_sign) goto catch_error;
+  if (self->input_buffer_idx >= MAX_BUFFER_LENGTH) {
+    // |input_buffer| is full. Buffers this long are not supported.
+    status = SV_NOT_SUPPORTED;
+    goto catch_error;
+  }
+
+  if (!self->input_buffer[self->input_buffer_idx]) {
+    self->input_buffer[self->input_buffer_idx] = calloc(1, signature_info->hash_size);
+    if (!self->input_buffer[self->input_buffer_idx]) {
+      // Failed in memory allocation.
+      status = SV_MEMORY;
+      goto catch_error;
+    }
     self->hash_size = signature_info->hash_size;
   }
 
   // Currently a fixed |hash_size| throughout the session is assumed.
   // TODO: Should we allow to change the hash_size in runtime?
-  if (signature_info->hash_size != self->hash_size) goto catch_error;
+  if (signature_info->hash_size != self->hash_size) {
+    status = SV_NOT_SUPPORTED;
+    goto catch_error;
+  }
 
   // Copy the |hash| ready for signing.
-  memcpy(self->hash_to_sign, signature_info->hash, signature_info->hash_size);
+  memcpy(
+      self->input_buffer[self->input_buffer_idx], signature_info->hash, signature_info->hash_size);
+  self->input_buffer_idx++;
 
   status = SV_OK;
-  self->plugin_state = THREADED_SIGNING_HAS_HASH_TO_SIGN;
 
 catch_error:
-  if (status == SV_UNKNOWN_FAILURE) {
-    // Failed in memory allocation. Free all memory and set status to SV_MEMORY.
+  if (status != SV_OK) {
+    // Failed in memory allocation or failure due to full buffer. Free all memory.
     sv_threaded_plugin_reset(self);
```

Comment:
I'm not sure it is a correct operation to clear all memory for all types of failure.
If we return a failure the ongoing SEI will be deleted and memory freed. But there may be others waiting for a signature, and those might wait for ever.
I suggest that we only reset if the thread has stopped.